cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -185,54 +183,50 @@ public Watch createWatch(
   /**
    * {@inheritDoc}
    */
-  @Override public Collection<T> groups() {
-    readLock.lock();
-
-    try {
-      return ImmutableSet.copyOf(statuses.keySet());
-    } finally {
-      readLock.unlock();
-    }
+  @Override
+  public Collection<T> groups() {
+    return ImmutableSet.copyOf(statuses.keySet());
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void setSnapshot(T group, Snapshot snapshot) {
+    // we take a writeLock to prevent watches from being created while we update the snapshot
+    CacheStatusInfo<T> status;
     writeLock.lock();
-
     try {
       // Update the existing snapshot entry.
       snapshots.put(group, snapshot);
+      status = statuses.get(group);
+    } finally {
+      writeLock.unlock();
+    }
 
-      CacheStatusInfo<T> status = statuses.get(group);
+    if (status == null) {
+      return;
+    }
 
-      if (status == null) {
-        return;
-      }
+    status.watchesRemoveIf((id, watch) -> {
```

Comment:
This is the hardest part, but I think you nailed it.

Let's assume this situation:
There is `setSnapshot(group, snapshot1)` and `setSnapshot(group, snapshot2)` calls concurrently called.
- `setSnapshot(group, snapshot1)` fires first, it sets the snapshot in `snapshots` and it starts iterating over watches in `status`
- `setSnapshot(group, snapshot2)` then fires, overrides snapshot in `snapshots` and starts iterating over watches in `status` competing with another call who will send which one. 

If the old snapshot is sent for a watch, then it will immediately create another one with `createWatch`, see that the version is old and immediately respond with a new one, because `createWatch` sees the new snapshot all the time.

So the number of calls to `createWatch` stay the same before and after this change. Am I right?

Reply:
Yup, that is exactly the case, the most problematic thing would be if `createWatch` were not seeing the latest version, because then, watches could be left unanswered.

But you've raised something i hadn't thought about, if two `setSnapshot` fire concurrently, they could overlap themselves while answering watches, this could probably allow for a watch to be responded twice. We could probably put another lock around to prevent two concurrent `setSnapshots`.
This scenario raises a couple of other questions in my head, right now a Watch is intended to be created and responded a single time, but there is nothing actually validating this, the semantics are a little weird. It would be good to place this kind of checks directly on the watches, or make watches more durable, which is something i find more interesting.

Reply:
I think that making `setSnapshot` synchronized makes sense, but what about lock per `group` so running `setSnapshot` for `group1` and `setSnapshot` for `group2` can run concurrently. Watch is only for one group, so they won't overlap.
What do you think about it?

Reply:
this is a feature i'd like to have, but it isn't something that was supported before, maybe we could leave it for another PR?

Reply:
Sure, I think we can do it in a separate PR.